File: bindings/python/examples/advanced_market_data.py
#!/usr/bin/env python3
"""
Advanced Market Data Operations

This example demonstrates advanced SDK features:
1. Using helper methods for easier asset/pair lookup
2. Converting between display and atomic units
3. Amount validation
4. Caching strategies
5. Getting quotes by ticker symbols
"""

import logging
import os

from kaleidoswap import (
    KaleidoClient, 
    KaleidoConfig, 
    Asset, 
    KaleidoError, 
    AssetNotFoundError,
    TradingPairNotFoundError,
    ValidationError
)

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

# Configuration
API_URL = os.getenv("KALEIDO_API_URL", "http://localhost:8000")


def main():
    """Demonstrate advanced market data operations."""

    # Initialize client
    config = KaleidoConfig(base_url=API_URL)
    client = KaleidoClient(config)

    # Example 1: Find assets by ticker
    logger.info("🔍 Example 1: Finding assets by ticker...")
    
    try:
        btc_asset = client.get_asset_by_ticker("BTC")
        if btc_asset:
            logger.info("  Found BTC:")
            logger.info(f"    Ticker: {btc_asset.ticker}")
            logger.info(f"    Precision: {btc_asset.precision}")
            logger.info(f"    Name: {btc_asset.name}")
    except AssetNotFoundError:
        logger.warning("  Asset 'BTC' not found.")
        btc_asset = None
    except KaleidoError as e:
        logger.warning(f"  Error finding BTC: {e}")
        btc_asset = None
    except Exception as e:
        logger.warning(f"  Unexpected error: {e}")
        btc_asset = None

    # Example 2: Find trading pair by ticker
    logger.info("\n📊 Example 2: Finding trading pairs by ticker...")
    
    try:
        pairs = client.list_pairs()
        if pairs:
            first_pair = pairs[0]
            pair_ticker = f"{first_pair.base.ticker}/{first_pair.quote.ticker}"
            logger.info(f"  Looking for pair: {pair_ticker}")

            pair = client.get_pair_by_ticker(pair_ticker)
            if pair:
                logger.info("  ✅ Found pair:")
                logger.info(f"    Base: {pair.base.ticker}")
                logger.info(f"    Quote: {pair.quote.ticker}")
    except TradingPairNotFoundError:
        logger.warning(f"  Trading pair {pair_ticker} not found.")
    except KaleidoError as e:
        logger.warning(f"  Could not get pairs: {e}")
    except Exception as e:
        logger.warning(f"  Unexpected error: {e}")
    # Example 3: Amount conversion
    logger.info("\n💱 Example 3: Converting between display and atomic units...")

    if btc_asset:
        # Convert 1.5 BTC to atomic units (satoshis)
        display_amount = 1.5
        atomic_amount = client.to_raw(display_amount, btc_asset)
        logger.info(f"  {display_amount} BTC = {atomic_amount} satoshis")

        # Convert back
        converted_back = client.to_display(atomic_amount, btc_asset)
        logger.info(f"  {atomic_amount} satoshis = {converted_back} BTC")
        
        # Using convert_amount helper
        result = client.convert_amount(2.5, btc_asset, to="raw")
        logger.info(f"  2.5 BTC = {result} satoshis (via convert_amount)")
    else:
        # Demo with mock asset
        logger.info("  Using mock BTC asset for demo...")
        mock_btc = Asset(ticker="BTC", name="Bitcoin", precision=8)
        
        atomic = client.to_raw(1.0, mock_btc)
        logger.info(f"  1.0 BTC = {atomic} satoshis")
        
        display = client.to_display(100_000_000, mock_btc)
        logger.info(f"  100,000,000 satoshis = {display} BTC")

    # Example 4: Amount validation
    logger.info("\n✅ Example 4: Validating amounts...")

    try:
        test_amount = 0.01  # 0.01 BTC
        validation = client.validate_amount(test_amount, "BTC")

        logger.info(f"  Amount: {test_amount} BTC")
        logger.info(f"  Valid: {validation['valid']}")
        logger.info(f"  Raw Amount: {validation['raw_amount']}")
        logger.info(f"  Display Amount: {validation['display_amount']}")
        if validation["errors"]:
            logger.info(f"  Errors: {', '.join(validation['errors'])}")
    except ValidationError as e:
        logger.warning(f"  Validation error: {e}")
    except Exception as e:
        logger.warning(f"  Could not validate amount: {e}")
    # Example 5: Check if pair is tradeable
    logger.info("\n🔄 Example 5: Checking trade availability...")

    can_trade_btc_usdt = client.can_trade("BTC", "USDT")
    logger.info(f"  Can trade BTC/USDT: {can_trade_btc_usdt}")

    can_trade_fake = client.can_trade("FAKE", "COIN")
    logger.info(f"  Can trade FAKE/COIN: {can_trade_fake}")

    # Example 6: Get quote by pair ticker
    logger.info("\n💰 Example 6: Getting quote by pair ticker...")

    try:
        if pairs:
            first_pair = pairs[0]
            pair_ticker = f"{first_pair.base.ticker}/{first_pair.quote.ticker}"
            
            # Get base asset to check limits
            base_asset = client.get_asset_by_ticker(first_pair.base.ticker)
            
            # Find a valid amount from asset limits
            from_amount = 10_000_000  # Default 10M smallest units
            if base_asset and base_asset.endpoints:
                for endpoint in base_asset.endpoints:
                    if hasattr(endpoint, 'min_amount') and endpoint.min_amount:
                        from_amount = max(from_amount, endpoint.min_amount + endpoint.min_amount // 10)
                        break
            
            logger.info(f"  Using amount: {from_amount} for {pair_ticker}")
            
            quote = client.get_quote_by_pair(
                ticker=pair_ticker,
                from_amount=from_amount,
            )

            logger.info(f"  ✅ Quote for {pair_ticker}:")
            logger.info(f"    RFQ ID: {quote.rfq_id}")
            logger.info(f"    From: {quote.from_asset.amount} {quote.from_asset.ticker}")
            logger.info(f"    To: {quote.to_asset.amount} {quote.to_asset.ticker}")
    except ResourceNotFoundError as e:
        logger.warning(f"  Resource not found: {e}")
    except KaleidoError as e:
        logger.warning(f"  Error getting quote: {e}")
    except Exception as e:
        logger.warning(f"  Unexpected error: {e}")


    # Example 7: Cache management
    logger.info("\n🔄 Example 7: Cache management...")

    logger.info("  Getting assets...")
    assets1 = client.list_assets()
    logger.info(f"    Found {len(assets1)} assets")

    logger.info("  Refreshing cache...")
    client.refresh_cache()
    logger.info("    Cache cleared")

    assets2 = client.list_assets()
    logger.info(f"    Found {len(assets2)} assets after refresh")

    logger.info("\n✅ All advanced market data operations completed!")


if __name__ == "__main__":
    main()

File: bindings/typescript/examples/advanced-market-data.ts
/**
 * Advanced Market Data Operations
 * 
 * This example demonstrates advanced SDK features:
 * 1. Using helper methods for easier asset/pair lookup
 * 2. Converting between display and atomic units
 * 3. Amount validation
 * 4. Caching strategies
 * 5. Getting quotes by ticker symbols
 */

import { KaleidoClient, Asset } from '../src';

const API_URL = process.env.KALEIDO_API_URL || 'https://api.staging.kaleidoswap.com';

async function main() {
    console.log('🚀 Advanced Market Data Operations\n');

    // Initialize client
    const client = new KaleidoClient({ baseUrl: API_URL });

    // Example 1: Find assets by ticker
    console.log('🔍 Example 1: Finding assets by ticker...');
    try {
        const btcAsset = await client.getAssetByTicker('BTC');
        if (typeof btcAsset !== 'string') {
            console.log('  Found BTC:');
            console.log(`    Ticker: ${btcAsset.ticker}`);
            console.log(`    Precision: ${btcAsset.precision}`);
            console.log(`    Name: ${btcAsset.name}`);
        }
    } catch (e) {
        console.log(`  Could not find BTC: ${e}`);
    }

    // Example 2: Find trading pairs
    console.log('\n📊 Example 2: Finding trading pairs...');
    try {
        const pairs = await client.listPairs();
        const pairsList = typeof pairs === 'string' ? JSON.parse(pairs) : pairs;

        if (pairsList.length > 0) {
            const firstPair = pairsList[0];
            const pairTicker = `${firstPair.baseAsset}/${firstPair.quoteAsset}`;
            console.log(`  Found pair: ${pairTicker}`);
            console.log(`    Base: ${firstPair.baseAsset}`);
            console.log(`    Quote: ${firstPair.quoteAsset}`);
        }
    } catch (e) {
        console.log(`  Could not get pairs: ${e}`);
    }

    // Example 3: Amount conversion
    console.log('\n💱 Example 3: Converting between display and atomic units...');
    try {
        // Create a mock BTC asset for demo
        const mockBtc: Asset = {
            assetId: 'bitcoin',
            ticker: 'BTC',
            name: 'Bitcoin',
            precision: 8,
            issuedSupply: 0,
            timestamp: 0,
            addedAt: 0,
            balance: { settled: 0, future: 0, spendable: 0, offchainOutbound: 0, offchainInbound: 0 },
        };

        // Convert 1.5 BTC to satoshis using object
        const raw = await client.toRaw(1.5, mockBtc);
        console.log(`  1.5 BTC = ${raw} satoshis`);

        // Convert back
        const display = await client.toDisplay(raw, mockBtc);
        console.log(`  ${raw} satoshis = ${display} BTC`);

        // Using convertAmount helper
        const raw2 = await client.convertAmount(2.5, mockBtc, 'raw');
        console.log(`  2.5 BTC = ${raw2} satoshis (via convertAmount)`);
    } catch (e) {
        console.log(`  Conversion error: ${e}`);
    }

    // Example 4: Amount validation
    console.log('\n✅ Example 4: Validating amounts...');
    try {
        const validation = await client.validateAmount(0.01, 'BTC');
        console.log(`  Amount: 0.01 BTC`);
        console.log(`  Valid: ${validation.valid}`);
        console.log(`  Raw Amount: ${validation.rawAmount}`);
        console.log(`  Errors: ${validation.errors.length > 0 ? validation.errors.join(', ') : 'None'}`);
    } catch (e) {
        console.log(`  Could not validate: ${e}`);
    }

    // Example 5: Check if pair is tradeable
    console.log('\n🔄 Example 5: Checking trade availability...');
    const canTradeBtcUsdt = await client.canTrade('BTC', 'USDT');
    console.log(`  Can trade BTC/USDT: ${canTradeBtcUsdt}`);

    const canTradeFake = await client.canTrade('FAKE', 'COIN');
    console.log(`  Can trade FAKE/COIN: ${canTradeFake}`);

    // Example 6: Cache management
    console.log('\n🔄 Example 6: Cache management...');
    console.log('  Getting assets...');
    const assets1 = await client.listAssets();
    const assetsList = typeof assets1 === 'string' ? JSON.parse(assets1) : assets1;
    console.log(`    Found ${assetsList.length} assets`);

    console.log('  Refreshing cache...');
    client.refreshCache();
    console.log('    Cache cleared');

    const assets2 = await client.listAssets();
    const assetsList2 = typeof assets2 === 'string' ? JSON.parse(assets2) : assets2;
    console.log(`    Found ${assetsList2.length} assets after refresh`);

    console.log('\n✅ All advanced market data operations completed!');
}

main().catch(console.error);
